File: python-chatdownloader/src/metrics/emote.py
"""
The emote metrics
"""
import logging
from dataclasses import dataclass

import requests
from _types import ChatMessageFragment, Comment
from consts import VED_CH_ID

from .abstractmetric import AbstractMetric

WEIGHT_EMOTES = 0.02
SEVEN_TV_URL = f"https://7tv.io/v3/users/twitch/{VED_CH_ID}"


@dataclass
class SevenTVEmote:
    name: str
    # NOTE: Will need this in the future, but it's only here now
    # because I'm lazy to rediscover where to get the image url
    emote_url: str


class Emote(AbstractMetric):
    """
    The emote metric.
    """

    def __init__(self):
        self.seventv_emotes = self.__get_all_7tv_emotes()
        self.seventv_lookup = {emote.name for emote in self.seventv_emotes}

    @staticmethod
    def __get_all_7tv_emotes() -> list[SevenTVEmote]:
        logging.info('Getting the 7TV channel emotes')
        response = requests.get(SEVEN_TV_URL, timeout=5)
        if not response.ok:
            logging.warning('Cannot get 7tv emotes')
            return []

        resp_body = response.json()
        ret_val = []
        try:
            raw_emotes = resp_body['emote_set']['emotes']
            for raw_emote in raw_emotes:
                host_url = raw_emote['data']['host']['url']
                filename = max(filter(lambda x: x['name'].endswith('.webp'),
                                      raw_emote['data']['host']['files']),
                               key=lambda emote: emote['width'])
                ret_val.append(SevenTVEmote(
                    name=raw_emote['name'],
                    emote_url=f'https://{host_url}/{filename}'
                ))
        except KeyError as e:
            logging.error('Cannot access the required keys to get the emotes',
                          exc_info=e)
            return []

        logging.info('Got %d 7tv emotes', len(ret_val))
        return ret_val

    def __count_7tv_emotes_in_fragment(self, fragment: ChatMessageFragment):
        count = 0
        for word in fragment.text.split(' '):
            if word in self.seventv_lookup:
                count += 1
        logging.debug('Found %d number of 7TV emotes in %s',
                      count, fragment.text)
        return count

    @classmethod
    def can_parallelize(cls) -> bool:
        return False

    @classmethod
    def get_name(cls) -> str:
        return 'emote'

    def get_metric(self, comment: Comment,
                   sequence_no: int) -> dict[str, float]:
        return self._shortcut_for_this_comment_user(
            comment,
            sum(int(fragment.emoticon is not None) +
                self.__count_7tv_emotes_in_fragment(fragment)
                for fragment in comment.message.fragments) * WEIGHT_EMOTES)

File: chatdownloader/src/metrics/emote.rs
/*
The emote metric
*/

use std::collections::{HashMap, HashSet};

use log::{info, debug};
use serde::Deserialize;
use lazy_static::lazy_static;

use crate::_types::twitchtypes::{ChatMessageFragment, Comment};
use crate::_constants::VED_CH_ID;

use super::metrictrait::AbstractMetric;

const WEIGHT_EMOTES: f32 = 0.02;

lazy_static! {
    static ref SEVEN_TV_URL: String = format!("https://7tv.io/v3/users/twitch/{}", VED_CH_ID);
}

#[derive(Deserialize)]
#[allow(dead_code)]
struct SevenTVEmote {
    name: String,
    emote_url: String,
}

#[allow(dead_code)]
pub struct Emote {
    seventv_emotes: Vec<SevenTVEmote>,
    seventv_lookup: HashSet<String>,
}

impl Emote {
    fn count_7tv_emotes_in_fragment(&self, fragment: &ChatMessageFragment) -> usize {
        let mut count = 0;
        for word in fragment.text.split(' ') {
            if self.seventv_lookup.contains(word) {
                count += 1;
            }
        }
        debug!("Found {} number of 7TV emotes in {}", count, fragment.text);
        count
    }
}

impl AbstractMetric for Emote {
    async fn new() -> Self {
        info!("Getting the 7TV channel emotes");
        let response = reqwest::get(SEVEN_TV_URL.clone()).await;
        if response.is_err() {
            info!("Cannot get 7tv emotes");
            return Self {
                seventv_emotes: Vec::new(),
                seventv_lookup: HashSet::new(),
            };
        }

        let resp_body: serde_json::Value = response.unwrap().json().await.unwrap();
        let mut ret_val = Vec::new();
        if let Some(raw_emotes) = resp_body["emote_set"]["emotes"].as_array() {
            for raw_emote in raw_emotes {
                let host_url = raw_emote["data"]["host"]["url"].as_str().unwrap();
                let filename = raw_emote["data"]["host"]["files"]
                    .as_array()
                    .unwrap()
                    .iter()
                    .filter(|emote| emote["name"].as_str().unwrap().ends_with(".webp"))
                    .max_by_key(|emote| emote["width"].as_i64().unwrap())
                    .unwrap();
                ret_val.push(SevenTVEmote {
                    name: raw_emote["name"].as_str().unwrap().to_owned(),
                    emote_url: format!("https://{}/{}", host_url, filename["name"]),
                });
            }
        } else {
            info!("Cannot access the required keys to get the emotes");
        }

        debug!("Got {} 7tv emotes", ret_val.len());
        let seventv_lookup: HashSet<String> = ret_val.iter().map(|emote| emote.name.clone()).collect();
        Self {
            seventv_emotes: ret_val,
            seventv_lookup,
        }
    }

    fn can_parallelize(&self) -> bool {
        false
    }

    fn get_name(&self) -> String {
        String::from("emote")
    }

    fn get_metric(&mut self, comment: Comment, _sequence_no: u32) -> HashMap<String, f32> {
        let metric = comment
            .message
            .fragments
            .iter()
            .map(|fragment| {
                (fragment.emoticon.is_some() as u16 as f32
                    + self.count_7tv_emotes_in_fragment(fragment) as f32)
                    * WEIGHT_EMOTES
            })
            .sum();
        let mut result = HashMap::new();
        result.insert("emote".to_owned(), metric);
        result
    }
}